src/main/java/org/veupathdb/service/eda/ds/core/AbstractPlugin.java

```diff
@@ -611,4 +659,155 @@ public String getRBinListAsString(List<String> labels) {
+    return dataSpecsWithStudyDependentVocabs;
+  }
+
+  public List<DynamicDataSpecImpl> findVariableSpecsWithStudyDependentVocabs(Map<String, VariableSpec> varSpecs) {
+    Map<String, DynamicDataSpecImpl> dataSpecs = varMapToDynamicDataMap(varSpecs);
+
+    return findDataSpecsWithStudyDependentVocabs(dataSpecs);
+  }
+
+  public List<DynamicDataSpecImpl> findCollectionSpecsWithStudyDependentVocabs(Map<String, CollectionSpec> collectionSpecs) {
+    Map<String, DynamicDataSpecImpl> dataSpecs = collectionMapToDynamicDataMap(collectionSpecs);
+
+    return findDataSpecsWithStudyDependentVocabs(dataSpecs);
+  }
+
+  public String getDynamicDataSpecEntityId(DynamicDataSpecImpl dataSpec) {
+    if (dataSpec.isCollectionSpec()) {
+      return dataSpec.getCollectionSpec().getEntityId();
+    } else if (dataSpec.isVariableSpec()) {
+      return dataSpec.getVariableSpec().getEntityId();
+    } else {
+      return null;
+    }
+  }
+
+  public String getDynamicDataSpecId(DynamicDataSpecImpl dataSpec) {
+    if (dataSpec.isCollectionSpec()) {
+      return dataSpec.getCollectionSpec().getCollectionId();
+    } else if (dataSpec.isVariableSpec()) {
+      return dataSpec.getVariableSpec().getVariableId();
+    } else {
+      return null;
+    }
+  }
+
+  public String getRMegastudyAsString(String compressedDataHandle, Map<String, DynamicDataSpecImpl> dataSpecs) {
+    PluginUtil util = getUtil();
+
+    // find and validate variables w study specific vocabs
+    List<DynamicDataSpecImpl> dataSpecsWithStudyDependentVocabs = findDataSpecsWithStudyDependentVocabs(dataSpecs);
+    List<String> entities = dataSpecsWithStudyDependentVocabs.stream().map(data -> getDynamicDataSpecEntityId(data)).toList();
+    String studyVocabsAsRString = getRStudyVocabsAsString(dataSpecsWithStudyDependentVocabs);
+
+    // get ancestor ids, can use first var bc we validate there is at least one and they all have the same entity
+    String entityId = entities.get(0);
+    String ancestorIdsAsRString = util.getEntityAncestorsAsRVectorString(entityId, _referenceMetadata);
+
+    String megastudyAsRString = "veupathUtils::Megastudy(" + 
+                                 "data = " + compressedDataHandle + "," +
+                                 "ancestorIdColumns = " + ancestorIdsAsRString + "," +
+                                 "studySpecificVocabularies = " + studyVocabsAsRString + ")";
+
+    return megastudyAsRString;
+  }
+
+  public String getRInputDataWithImputedZeroesAsString(String compressedDataHandle, Map<String, DynamicDataSpecImpl> dataSpecs) {
+    boolean validRequest = validateImputeZeroesRequest(dataSpecs);
+    
+    if (validRequest) {
+      String megastudyData = getRMegastudyAsString(compressedDataHandle, dataSpecs);
+      // this feels hacky, but bc this has to be called from the plugin and the plugin already has variables in R... im assuming its present.
+      return "veupathUtils::getDTWithImputedZeroes(" + megastudyData + ", variables)";
```

Comment:
Maybe "variables" could be a parameter so we don't need outside knowledge to understand this method.